src/Language/Haskell/TH/Datatype.hs

```diff
@@ -178,16 +179,16 @@ import           Data.Monoid (Monoid(..))
 --   Then in the 'DatatypeInfo' for @F@'s data instance, we would have:
 --
 --   @
---   'datatypeVars'      = [ 'KindedTV' c 'StarT'
---                         , 'KindedTV' f 'StarT'
---                         , 'KindedTV' x 'StarT' ]
+--   'datatypeVars'      = [ 'KindedTV' c () 'StarT'
+--                         , 'KindedTV' f () 'StarT'
+--                         , 'KindedTV' x () 'StarT' ]
 --   'datatypeInstTypes' = [ 'AppT' ('ConT' ''Maybe) ('VarT' c)
 --                         , 'AppT' ('VarT' f) ('VarT' x) ]
 --   @
 data DatatypeInfo = DatatypeInfo
   { datatypeContext   :: Cxt               -- ^ Data type context (deprecated)
   , datatypeName      :: Name              -- ^ Type constructor
-  , datatypeVars      :: [TyVarBndr]       -- ^ Type parameters
+  , datatypeVars      :: [TyVarBndrUnit]   -- ^ Type parameters
```

Comment:
This is breaking change, so yes, major bump is fine. Personally I have bounds on `template-haskell`, so if I need to relax / update  `th-abstraction` bounds at the same time, it's not biggie.